sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -143,27 +146,18 @@ private JdbcIO() {}
     @Nullable abstract String getPassword();
     @Nullable abstract DataSource getDataSource();
 
-    /** Configuration using a {@link Serializable} {@link DataSource}. */
-    public static DataSourceConfiguration create(DataSource dataSource) {
-      checkNotNull(dataSource, "dataSource");
-      checkArgument(dataSource instanceof Serializable, "dataSource must be Serializable");
-      return new AutoValue_JdbcIO_DataSourceConfiguration(null, null, null, null, dataSource);
+    static Builder builder() {
```

Comment:
This must be public, or users won't be able to use it. The Builder class and all its methods also have to be, if that's the user-facing API you want to expose.

However, looking at the test code now, I'm not sure exposing the Builder is the right thing to do - it feels like exposing too much AutoValue API, whereas I'd prefer if AutoValue was used as an implementation detail, otherwise hidden from the user.

There's 2 use cases:

* Create from data source, and optional username/password
* Create from driver class name, url, and optional username/password
(anything else I'm missing?)

Perhaps the best API for this is 2 create() methods - `create(DataSource dataSource)` and `create(String driverClassName, String url)`, plus methods `withUsername`/`withPassword`. Without exposing the Builder class, nor the setBlah()/builder()/build() methods.

user code should look like:
* `JdbcIO.DataSourceConfiguration.create(dataSource)`
* `JdbcIO.DataSourceConfiguration.create(dataSource).withUsername("sa").withPassword("sa")`
* `JdbcIO.DataSourceConfiguration.create(driverClassName, url).withUsername("sa").withPassword("sa")`
etc.

This way you are safe from accidentally creating a DataSourceConfiguration that has both a dataSource and a driverClassName.